File: ios-native/MGCloudSync.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(MGCloudSync, RCTEventEmitter)

RCT_EXTERN_METHOD(getValue:(NSString *)key
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setValue:(NSString *)key
                  value:(NSString *)value
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(removeValue:(NSString *)key
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getAllKeys:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios-native/MGCloudSync.swift
import Foundation

@objc(MGCloudSync)
class MGCloudSync: RCTEventEmitter {
  private let store = NSUbiquitousKeyValueStore.default

  override init() {
    super.init()
    store.synchronize()
    NotificationCenter.default.addObserver(
      self,
      selector: #selector(externalChange(_:)),
      name: NSUbiquitousKeyValueStore.didChangeExternallyNotification,
      object: store
    )
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }

  @objc func externalChange(_ notification: Notification) {
    guard let keys = notification.userInfo?[NSUbiquitousKeyValueStoreChangedKeysKey] as? [String] else { return }
    sendEvent(withName: "MGCloudSyncExternalChange", body: ["keys": keys])
  }

  override func supportedEvents() -> [String]! {
    return ["MGCloudSyncExternalChange"]
  }

  @objc func getValue(
    _ key: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(store.string(forKey: key))
  }

  @objc func setValue(
    _ key: String,
    value: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    store.set(value, forKey: key)
    store.synchronize()
    resolve(nil)
  }

  @objc func removeValue(
    _ key: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    store.removeObject(forKey: key)
    store.synchronize()
    resolve(nil)
  }

  @objc func getAllKeys(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(Array(store.dictionaryRepresentation.keys))
  }

  override class func requiresMainQueueSetup() -> Bool {
    return false
  }
}
